### ISSUE TITLE
`eslint-` comments cause `lines-around-comment` rule to error

### ISSUE THREAD

**Issue by mathieumg** [eslint/eslint#4345]
1. **The version of ESLint you are using**
   
   1.8.0
2. **What you did (the source code and ESLint configuration)**
   
   ``` js
   /* eslint-disable no-underscore-dangle */
   this._values = values;
   this._values2 = true;
   /* eslint-enable no-underscore-dangle */
   ```
   
   with the following configuration:
   
   ``` js
   "lines-around-comment": [2, {
     "allowArrayStart": true,
     "allowBlockStart": true,
     "allowObjectStart": true,
     "beforeBlockComment": true,
     "beforeLineComment": true
   }],
   ```
3. **The actual ESLint output complete with numbers**
   
   `error  Expected line before comment  lines-around-comment`
4. **What you expected to happen instead**
   
   No error. I don't think `lines-around-comment` should be enforced for `eslint-enable`/`eslint-disable` type comments. This follows with the fact that `no-inline-comments` ignores `eslint-disable-line` comments even when it's enabled.
##

**Comment by eslintbot** [comment]
Thanks for the issue! If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

**Comment by gyandeeps** [comment]
Why not do it like this
```js
this._values = values; // eslint-disable-line no-underscore-dangle
```

**Comment by mathieumg** [comment]
I have multiple lines in-between, I'll edit my example to be more representative of the original code.

**Comment by mathieumg** [comment]
I do have another area in the same code where I use `// eslint-disable-line no-underscore-dangle` already, and it works correctly, so there's that!

**Comment by gyandeeps** [comment]
@eslint/eslint-team thoughts?

**Comment by nzakas** [comment]
I don't know - we seem to always run into this question of whether configuration comments should be treated as comments or not. It seems like something we should answer globally for every rule rather than attacking it in just one rule.

**Comment by mathieumg** [comment]
> we should answer globally for every rule rather than attacking it in just one rule

I second that! I mentioned the `no-inline-comments` that ignores the `eslint-disable-line` because I remembered it had gotten "fixed", but I imagine there may be more rules that correspond to that scenario. I'd be curious which other there are, I didn't find any by quickly glancing at the rule list. Perhaps there would be `no-warning-comments` but writing:

```js
// TODO: Implement xyz. // eslint-enable-line no-warning-comments
```

seems odd to me. I don't use that rule, but perhaps that would be what people that do use it prefer.

**Comment by mathieumg** [comment]
I know you're all swamped with the work on 2.0 (great job so far, by the way!), is there any way I can help this progress further? Thanks.

**Comment by nzakas** [comment]
I think what we need here is to get a consensus that we should basically exempt all ESLint comments from all checks. @eslint/eslint-team thoughts?

After that, we'd need to see if there's an easy way to filter those comments out. If you'd like to explore that in a PR, that would be helpful.

**Comment by ilyavolodin** [comment]
I think ignoring configuration comments is a good idea. They are only there to satisfy ESLint, and not part of the core code. So they shouldn't be subject to linting.

**Comment by platinumazure** [comment]
:+1: on filtering out all configuration comments from linting if we can.

**Comment by nzakas** [comment]
It looks like we all agreed this was a good idea. Anyone want to take a stab at implementing it?

**Comment by BYK** [comment]
I am up for the task. I'll add an option to `lines-around-comment`, `ignoreDirectives` or `ignoreESLintDirectives` which would ignore all ESLint directives (not just `eslint-disable` and `eslint-enable`).

Sounds like a good plan?

**Comment by nzakas** [comment]
@BYK I think the suggestion was instead to do this in the core (labels were wrong) instead of needing to manually do this in every rule.

**Comment by BYK** [comment]
@nzakas - oh right, I didn't read the whole conversation, sorry. Alright I'll make ESLint not trigger comment visitors for directives then, makes sense?

**Comment by alberto** [comment]
Mmm, this is a bit tricky. I am not sure this would be desired either:

```
"lines-around-comment": [2, {
  "beforeBlockComment": true
}],
```

```
foo();
/* eslint-disable no-underscore-dangle */
this._values = values;
this._values2 = true;
/* eslint-enable no-underscore-dangle */
```

**Comment by nzakas** [comment]
@alberto what are you saying would happen there?

@BYK I think you'd also need to remove them from `trailingComments` and `leadingComments` (probably fine to do so at the time of calling `context.getComments()`).

**Comment by BYK** [comment]
@nzakas - thanks for the pointer. Will look there.

**Comment by alberto** [comment]
@nzakas that it would not warn on the second line with the proposed changes. I understand why you wouldn't want lines inside the comments regardless of the rule configuration. I am not so sure about the outside.

**Comment by nzakas** [comment]
@alberto by "it" do you mean the `lines-around-comment` rule? If so, I believe that's the intended behavior based on this issue. It probably wouldn't make sense to partially apply a rule to these comments. 

Although, this is an interesting point, and this change is probably breaking so deserves closer consideration.

**Comment by alberto** [comment]
@nzakas Yes. I think I would like to write it like:

```
foo();

/* eslint-disable no-underscore-dangle */
this._values = values;
this._values2 = true;
/* eslint-enable no-underscore-dangle */

bar();
```

Note this could also trigger `no-multiple-empty-lines {max: 1}`  if eslint comments are removed completely.

**Comment by nzakas** [comment]
Well, this is the problem: either these comments are treated as comments or as empty spaces, either way there's going to be something "wrong". So we can either note make this change and accept that people will think these comments shouldn't count, or we leave them as-is and accept that people will think these comments should count. 

Or we add an option to every comment rule that lets people choose.

**Comment by jthomerson** [comment]
Could the `eslint-disable` and `eslint-enable` directives be made to just work with single-line (as opposed to block) comments? That seems like it would (mostly) solve the issue as people generally don't require lines before/after single-line comments like they do block comments. Or, potentially an `ignoreRulePatterns` config option could be added that allowed you to ignore the `lines-around-comment` rules for comments that match a certain pattern. The documentation could give this example: `ignoreRulePatterns: [ /^ eslint-(dis|en)able .*/ ];`

**Comment by nzakas** [comment]
@jthomerson that's not a good solution because we can't have everything change the `eslint-disable` comments to single line comments. That's very unfair to current users.

**Comment by jthomerson** [comment]
@nzakas I wasn't saying that single-line comments (`//`) had to replace existing block comments (`/* ... */`). I'm just saying that I don't see a reason that eslint should not allow for either style of comment. `eslint-enable` and `eslint-disable` comments, in particular, are hardly ever more than one line, so why should eslint force those to use block comment style, which is typically only used for multi-line comments?

**Comment by nzakas** [comment]
@jthomerson ah I see, thanks for clarifying. The reason we don't allow single-line comments is because it's not clear when they would apply and when they wouldn't. The single line comments we do allow (`// eslint-disable-line` and `// eslint-disable-next-line`) make it very clear what they apply to. 

In any event, this topic isn't really relevant to the issue at hand. If you'd like to chat more about it, please stop by the chatroom.

**Comment by jthomerson** [comment]
@nzakas thanks for the extra explanation. However, I'm not sure how the block comments help clarify where the rules apply or don't apply any more than the single-line comments since the block comments are not "wrapping" the code they apply to anyway - they're merely indicating the starting position of a different rule.

Also, I don't see how you can say that my suggestion is not relevant to the issue at hand. The reporter has the following code:

```
/* eslint-disable no-underscore-dangle */
this._values = values;
this._values2 = true;
/* eslint-enable no-underscore-dangle */
```

That, along with a very standard `lines-around-comment` rule that requires one line `beforeBlockComment`, means that he has to do this:

```
/* eslint-disable no-underscore-dangle */
this._values = values;
this._values2 = true;

/* eslint-enable no-underscore-dangle */
```

I'm suggesting this be supported:

```
// eslint-disable no-underscore-dangle
this._values = values;
this._values2 = true;
// eslint-enable no-underscore-dangle
```

That seems relevant to this discussion - IMHO. I'm up against the same issue as the reporter and would like to see a solution to it, and using single-line comments seems like a valid solution to me.

Thanks again. We love the work done on eslint!

**Comment by nzakas** [comment]
@jthomerson I feel like I've already explained a couple of times why your suggestion isn't relevant to the topic at hand. You're suggesting *another* way to disable rules inline, which is a feature request that does not solve the problem for the *current* way to disable rules inline.

So I'd like to ask, once again, that we stick to the original problem of how to deal with what we currently support for configuration comments.

**Comment by IanVS** [comment]
Personally, I think that triggering an error for `no-multiple-empty-lines` would be _way_ more confusing than the current error of `lines-around-comment`.  Yeah, it might be irritating the way it is now, but I would see it as a bug to trigger a `no-multiple-empty-lines` error.  The only acceptable solution I can see is if we could completely treat the line like it doesn't exist, which I'm guessing is not so easy.

**Comment by nzakas** [comment]
I feel like we just keep going round and round on this, so maybe the best solution is just not to make any changes. What we had now may be a bit weird, but it seems like any potential change to that behavior has some other weird side effect or implication.

**Comment by nzakas** [comment]
Per [TSC meeting](https://github.com/eslint/tsc-meetings/blob/master/notes/2016/07-Jul-2016.md#httpsgithubcomeslinteslintissues4345), we decided that globally excluding comments has too many problems and we shouldn't proceed down that path. We do think it's useful to be able to exclude configuration comments from `lines-around-comments`, and have decided on adding an `ignorePattern` option to the rule that behaves the same as in https://github.com/eslint/eslint/issues/6077.

**Comment by ilyavolodin** [comment]
I think `ignorePattern` should be made into helper method, so it can be reused across multiple rules, if necessary.

**Comment by TheSavior** [comment]
This issue should probably be named a bit more generic since this isn't only affected by the enable and disable lines. 

eslint-env also causes this problem.

```
/* eslint-env mocha */
'use strict';
```

**Comment by kaicataldo** [comment]
I'll implement this

**Comment by nzakas** [comment]
@kaicataldo is this still on your radar?

**Comment by kaicataldo** [comment]
@nzakas It's in my backlog, but if someone else wants to take this on please feel free!

**Comment by alberto** [comment]
I'm working on this
